### PR TITLE
stat/combin: Implement functions for helping with Permutations

### LINKED ISSUE #1033: stat/combin: Add Permutations and PermutationsGenerator
Could consider Product as well (as in CartesianProduct)

### PR DESCRIPTION
**Pull request by btracey**
Fixes #1033.

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
I think this wants some tests, and probably also some examples to help out with the documentation.

**Comment by btracey** [comment]
This is a partial commit, because implemented the work and realized the issue.

Permutations are normally P(n, k), while often times in iteration you just want the iterations of the first k objects (like in the python itertools.permutations). I assume we want this to be the fully general form, so `IndexToPermutation(dst []int, idx, n, k int) int`? I'm pretty sure this is possible, because there's also a function `IndexToCombination`, and we can basically select the combination first and then the permutation. I'll make that the next PR.

This current implementation uses that `n = k`, and just uses the length of the appropriate slice to get both values. If we go to the more general version, would we rather keep inferring `k` from the length of the slice, and just require the user to specify `n`, or require the user to specify both? I lean toward the latter given that they're both integers and would be confusing from context otherwise.

**Comment by btracey** [comment]
Sorry @kortschak didn't get my note out fast enough!

File: stat/combin/combin.go [hunk 1]
```diff
@@ -185,6 +185,98 @@ func nextCombination(s []int, n, k int) {
...
+	p := 1
+	for i := n - k + 1; i <= n; i++ {
+		p *= i
+	}
+	return p
+}
+
+// factorial returns a!.
+func factorial(a int) int {
+	a := 1
```
Comment (kortschak) [inline stat/combin/combin.go]: `f`?

File: stat/combin/combin.go [hunk 2]
```diff
@@ -185,6 +185,98 @@ func nextCombination(s []int, n, k int) {
...
+// PermutationIndex panics if perm is not a permutation of the first len(perm)
+// non-negative integers.
+func PermutationIndex(perm []int) int {
+	// Note(btracey): This is an n^2 algorithm, but factorial increases
+	// very quickly (25! overflows int64) so this is not a problem in
+	// practice.
+
+	check := map[int]struct{}{}
+	for _, v := range perm {
+		if 0 <= v || v >= len(perm) {
```
Comment (vladimir-ch) [inline stat/combin/combin.go]: `0 <= v` should be `v < 0`?

Comment (kortschak) [reply stat/combin/combin.go]: Also, `len(perm) <= v`.

Comment (sbinet) [reply stat/combin/combin.go]: I guess an easier way to spell it might be:
```go
if !(0 <= v && v < len(perm)) {
```

Comment (kortschak) [reply stat/combin/combin.go]: We normally write value out of range as `v < 0 || len(perm) <= v`.

**Comment by btracey** [comment]
Closing in favor of #1076

**Pull request closed without merging**